datafusion/expr/src/tree_node/plan.rs

```diff
@@ -85,28 +85,25 @@ impl TreeNode for LogicalPlan {
         F: FnMut(Self) -> Result<Transformed<Self>>,
     {
         let old_children = self.inputs();
-        let t = old_children
+        let new_children = old_children
             .iter()
             .map(|&c| c.clone())
-            .map_till_continue_and_collect(f)?;
-        // TODO: Currently `assert_eq!(t.transformed, t2)` fails as
-        //  `t.transformed` quality comes from if the transformation closures fill the
-        //   field correctly.
-        //  Once we trust `t.transformed` we can remove the additional check in
-        //  `t2`.
-        let t2 = old_children
-            .into_iter()
-            .zip(t.data.iter())
-            .any(|(c1, c2)| c1 != c2);
+            .map(f)
+            .collect::<Result<Vec<_>>>()?;
 
-        // Propagate up `t.transformed` and `t.tnr` along with the node containing
-        // transformed children.
-        if t2 {
-            t.flat_map_data(|new_children| {
-                self.with_new_exprs(self.expressions(), new_children)
-            })
+        // if any changes made, make a new child
+        if old_children
+            .into_iter()
+            .zip(new_children.iter())
+            .any(|(c1, c2)| c1 != &c2.data)
+        {
+            self.with_new_exprs(
+                self.expressions(),
+                new_children.into_iter().map(|child| child.data).collect(),
+            )
+            .map(Transformed::yes)
```

Comment:
Why do you want to propogate up the `transformed` flag based on the detection result (`old_children.into_iter().zip(new_children.iter()).any(|(c1, c2)| c1 != &c2.data)`) and not based on if the transformation closure (`f`) applied on any children reported (propagated up)`transformed` true state? 

I mean I left similar TODOs in `DynTreeNode` and `ConcreteTreeNode` where the same conflic between the "detected" and "reported" transformed state can happen.
 
In my PR I always propagated up the trasnformed state from the closures despite the detection could be different. We can agree on to propagate the detected but then IMO we should change this in `DynTreeNode` and `ConcreteTreeNode` as well.

Reply:
Relevant topic in the main PR: https://github.com/apache/arrow-datafusion/pull/8891/files/e1a499ab51aea3623d1c2754f16fb70e6176d119#r1504355259

Reply:
I initially thought there may exist some rules such that it transforms the tree in a few iterations, but at the final stage, the final data is the same as the initial data. Therefore, it would be better to check the data's itself rather than trusting the transformed flag. However, now I think that the equality check at each step could be expensive, and I think we need to rely on transformed flag. Otherwise it would have no meaning to carry.

 I am converting it to propagate the transformed information up if any children is transformed.

Reply:
Honestly, my take on this general, user provided `transformed` flag in the transformation closure results is that it is pointless. In the current code on the `main` branch it doesn't work at all and so it isn't used anywhere for anything. This tells me that in most of the cases noone misses it. (You know I wanted to remove it entirelly first: https://github.com/apache/arrow-datafusion/pull/8835.)

Now the above doesn't mean that I can't imagine a case where we want to check if changes were made to the tree during a transformation. But I see other options without using the `transformed` flag.
- If it is a tree with `Arc` "links" and `enum` nodes like `LogicalPlan` we can compare the old cloned top level node with the new top level node we got back from transform to detect changes. This comparison is cheap (and similar to  the `old_children.into_iter().zip(new_children.iter()).any(|(c1, c2)| c1 != &c2.data)` above).
- If it is a tree with `Arc<dyn trait>` nodes (`Arc<DynTreeNode>` and `ConcreteTreeNode` trees fall into this category) then we can compare the old cloned top level node data ptr addres with the new top level node data prt address we got back from transform to detect changes. This is similar to how `with_new_children_if_necessary` works.
- If we have an `Expr` tree (with `Box` links) then it would be costly to clone the tree before the transformation so the only cheap way to detect if changes were made is to ask the API user to provide the information if they changed a node. But instead of requesting the transformed flag in all transformation closures from the API users, I would rather offer an new "transform with payload" like API (https://github.com/apache/arrow-datafusion/pull/8664) to let the user propagate arbitrary state during transformation. The `transformed` flag in the `Transformed` struct is nothing more than a special payload in that regard.

Reply:
The options you mentioned make sense, but instead of going through so much effort, we should use the already existing transform. I can't see any advantage of removing Transform either. I also use it a lot in the projection optimizer rule I'm currently working on. As far as I remember, there were also some ideas: https://github.com/apache/arrow-datafusion/pull/8835#issuecomment-1892570956.

I have sent a new commit related with these. Could you please take a look?

Reply:
Yes, I'm ok with keeping the flag, especially if you are going to use it in rules. But I have some concerns with that commit: https://github.com/peter-toth/arrow-datafusion/pull/1/files#r1505932547